NullPointerException when reading JsonManifest

I tried to push docker image to latest version of artipie registry:
```
docker push central.artipie.com/alpine:3.10
The push refers to repository [central.artipie.com/alpine]
77cae8ab23bf: Pushed 
[DEPRECATION NOTICE] registry v2 schema1 support will be removed in an upcoming release. Please contact admins of the central.artipie.com registry NOW to avoid future disruption. More information at https://docs.docker.com/registry/spec/deprecated-schema-v1/
received unexpected HTTP status: 500 Internal Server Error
```

But it fails with 500 error after push:
```
artipie         | [INFO] 2020-06-16 02:54:59,584 aws-java-sdk-NettyEventLoop-1-1 com.artipie.http.async.AsyncSlice - Failure: java.util.concurrent.CompletionException: java.lang.NullPointerException
artipie         | 	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
artipie         | 	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
artipie         | 	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source)
artipie         | 	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
artipie         | 	at java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)
artipie         | 	at io.reactivex.internal.observers.ConsumerSingleObserver.onSuccess(ConsumerSingleObserver.java:62)
artipie         | 	at io.reactivex.internal.operators.single.SingleMap$MapSingleObserver.onSuccess(SingleMap.java:64)
artipie         | 	at io.reactivex.internal.operators.single.SingleMap$MapSingleObserver.onSuccess(SingleMap.java:64)
artipie         | 	at io.reactivex.internal.operators.flowable.FlowableReduceSeedSingle$ReduceSeedObserver.onComplete(FlowableReduceSeedSingle.java:110)
artipie         | 	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:180)
artipie         | 	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.access$500(ResponseHandler.java:67)
artipie         | 	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$PublisherAdapter$1.onComplete(ResponseHandler.java:296)
artipie         | 	at com.typesafe.netty.HandlerPublisher.publishMessage(HandlerPublisher.java:362)
artipie         | 	at com.typesafe.netty.HandlerPublisher.flushBuffer(HandlerPublisher.java:304)
artipie         | 	at com.typesafe.netty.HandlerPublisher.receivedDemand(HandlerPublisher.java:258)
artipie         | 	at com.typesafe.netty.HandlerPublisher.access$200(HandlerPublisher.java:41)
artipie         | 	at com.typesafe.netty.HandlerPublisher$ChannelSubscription$1.run(HandlerPublisher.java:452)
artipie         | 	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
artipie         | 	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
artipie         | 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
artipie         | 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
artipie         | 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
artipie         | 	at java.base/java.lang.Thread.run(Unknown Source)
artipie         | Caused by: java.lang.NullPointerException
artipie         | 	at com.artipie.docker.manifest.JsonManifest.lambda$config$2(JsonManifest.java:91)
artipie         | 	... 21 more
```

--- a/pom.xml
+++ b/pom.xml
@@ -105,11 +105,6 @@ SOFTWARE.
       <scope>test</scope>
     </dependency>
     <dependency>
-      <groupId>org.slf4j</groupId>
-      <artifactId>slf4j-log4j12</artifactId>
-      <scope>test</scope>
-    </dependency>
-    <dependency>
       <groupId>log4j</groupId>
       <artifactId>log4j</artifactId>
       <scope>test</scope>

--- a/src/main/java/com/artipie/docker/asto/AstoBlobs.java
+++ b/src/main/java/com/artipie/docker/asto/AstoBlobs.java
@@ -25,6 +25,7 @@
 package com.artipie.docker.asto;
 
 import com.artipie.asto.Content;
+import com.artipie.asto.Remaining;
 import com.artipie.asto.Storage;
 import com.artipie.asto.fs.RxFile;
 import com.artipie.docker.Blob;
@@ -34,6 +35,7 @@ import io.reactivex.Completable;
 import io.reactivex.Flowable;
 import io.reactivex.Single;
 import java.io.IOException;
+import java.nio.ByteBuffer;
 import java.nio.channels.FileChannel;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -95,11 +97,8 @@ final class AstoBlobs implements BlobStore {
         return Flowable.fromPublisher(blob)
             .flatMapCompletable(
                 buf -> Completable.fromAction(
-                    () -> {
-                        while (buf.hasRemaining()) {
-                            out.write(buf);
-                        }
-                    })
+                    () -> out.write(ByteBuffer.wrap(new Remaining(buf, true).bytes()))
+                )
             )
             .doOnTerminate(out::close)
             .andThen(Single.just(out))

--- a/src/main/java/com/artipie/docker/asto/AstoManifests.java
+++ b/src/main/java/com/artipie/docker/asto/AstoManifests.java
@@ -77,20 +77,19 @@ public final class AstoManifests implements Manifests {
 
     @Override
     public CompletionStage<Manifest> put(final ManifestRef ref, final Content content) {
-        return new ByteBufPublisher(content).bytes()
-            .thenCompose(bytes -> this.blobs.put(new Content.From(bytes), new Digest.Sha256(bytes)))
-            .thenCompose(
-                blob -> {
-                    final Digest digest = blob.digest();
-                    return blob.content()
-                        .thenApply(source -> new JsonManifest(digest, source))
-                        .thenCompose(
-                            manifest -> this.validate(manifest)
-                                .thenCompose(nothing -> this.addManifestLinks(ref, digest))
-                                .thenApply(nothing -> manifest)
-                        );
-                }
-            );
+        return new ByteBufPublisher(content).bytes().thenCompose(
+            bytes -> {
+                final Content copy = new Content.From(bytes);
+                final Digest digest = new Digest.Sha256(bytes);
+                return this.blobs.put(copy, digest)
+                    .thenApply(blob -> new JsonManifest(digest, copy))
+                    .thenCompose(
+                        manifest -> this.validate(manifest)
+                            .thenCompose(nothing -> this.addManifestLinks(ref, digest))
+                            .thenApply(nothing -> manifest)
+                    );
+            }
+        );
     }
 
     @Override
@@ -102,6 +101,9 @@ public final class AstoManifests implements Manifests {
                         blobOpt -> blobOpt
                             .map(
                                 blob -> blob.content()
+                                    .thenApply(ByteBufPublisher::new)
+                                    .thenCompose(ByteBufPublisher::bytes)
+                                    .thenApply(Content.From::new)
                                     .<Manifest>thenApply(
                                         source -> new JsonManifest(blob.digest(), source)
                                     )

--- a/src/test/java/com/artipie/docker/http/S3CompatibilityITCase.java
+++ b/src/test/java/com/artipie/docker/http/S3CompatibilityITCase.java
@@ -24,22 +24,27 @@
 package com.artipie.docker.http;
 
 import com.adobe.testing.s3mock.junit5.S3MockExtension;
+import com.artipie.asto.Content;
 import com.artipie.asto.Storage;
 import com.artipie.asto.s3.S3Storage;
+import com.artipie.docker.Blob;
 import com.artipie.docker.Digest;
 import com.artipie.docker.Docker;
 import com.artipie.docker.RepoName;
 import com.artipie.docker.Upload;
 import com.artipie.docker.asto.AstoDocker;
 import com.artipie.docker.asto.BlobKey;
+import com.artipie.docker.ref.ManifestRef;
 import com.artipie.http.hm.RsHasStatus;
 import com.artipie.http.rq.RequestLine;
+import com.artipie.http.rs.Header;
 import com.artipie.http.rs.RsStatus;
 import io.reactivex.Flowable;
 import java.net.URI;
 import java.nio.ByteBuffer;
 import java.util.Collections;
 import java.util.UUID;
+import javax.json.Json;
 import org.hamcrest.MatcherAssert;
 import org.hamcrest.core.IsEqual;
 import org.junit.jupiter.api.BeforeEach;
@@ -56,8 +61,12 @@ import software.amazon.awssdk.services.s3.model.CreateBucketRequest;
  *
  * @since 0.3
  * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ * @todo #212:30min Remove S3CompatibilityITCase tests
+ *  Tests might be removed once issue resolved in ASTO: https://github.com/artipie/asto/issues/204
+ *  S3 mock dependency should be removed as well after that.
  */
-class UploadBlobS3ITCase {
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+class S3CompatibilityITCase {
 
     /**
      * Mock S3 server.
@@ -97,11 +106,11 @@ class UploadBlobS3ITCase {
         client.createBucket(CreateBucketRequest.builder().bucket(bucket).build()).join();
         this.storage = new S3Storage(client, bucket);
         this.docker = new AstoDocker(this.storage);
-        this.slice = new DockerSlice("/base", this.docker);
+        this.slice = new DockerSlice(this.docker);
     }
 
     @Test
-    void shouldUpload() {
+    void shouldUploadBlob() {
         final String name = "test";
         final Upload upload = this.docker.repo(new RepoName.Valid(name)).uploads()
             .start()
@@ -118,7 +127,7 @@ class UploadBlobS3ITCase {
                 new RequestLine(
                     "PUT",
                     String.format(
-                        "/base/v2/%s/blobs/uploads/%s?digest=%s",
+                        "/v2/%s/blobs/uploads/%s?digest=%s",
                         name,
                         upload.uuid(),
                         digest
@@ -134,5 +143,53 @@ class UploadBlobS3ITCase {
             this.storage.exists(new BlobKey(new Digest.FromString(digest))).join(),
             new IsEqual<>(true)
         );
+    }
+
+    @Test
+    void shouldPutManifest() {
+        MatcherAssert.assertThat(
+            this.slice.response(
+                new RequestLine("PUT", "/v2/test/manifests/1", "HTTP/1.1").toString(),
+                Collections.emptyList(),
+                this.manifest(new RepoName.Valid("test"))
+            ),
+            new RsHasStatus(RsStatus.CREATED)
+        );
+    }
+
+    @Test
+    void shouldGetManifest() {
+        final RepoName.Valid name = new RepoName.Valid("test");
+        this.docker.repo(name).manifests()
+            .put(new ManifestRef.FromString("2"), new Content.From(this.manifest(name)))
+            .toCompletableFuture().join();
+        MatcherAssert.assertThat(
+            this.slice.response(
+                new RequestLine("GET", "/v2/test/manifests/2", "HTTP/1.1").toString(),
+                Collections.singleton(
+                    new Header("Accept", "application/vnd.docker.distribution.manifest.v2+json")
+                ),
+                Flowable.empty()
+            ),
+            new RsHasStatus(RsStatus.OK)
+        );
+    }
+
+    private Flowable<ByteBuffer> manifest(final RepoName name) {
+        final byte[] content = "config".getBytes();
+        final Blob config = this.docker.repo(name).layers()
+            .put(new Content.From(content), new Digest.Sha256(content))
+            .toCompletableFuture().join();
+        final byte[] data = Json.createObjectBuilder()
+            .add("mediaType", "application/vnd.docker.distribution.manifest.v2+json")
+            .add(
+                "config",
+                Json.createObjectBuilder().add("digest", config.digest().string())
+            )
+            .add("layers", Json.createArrayBuilder())
+            .build()
+            .toString()
+            .getBytes();
+        return Flowable.just(ByteBuffer.wrap(data));
     }
 }

